File: lib/services/csv_export_service.dart
import 'dart:convert';
import 'package:csv/csv.dart';
import 'package:intl/intl.dart';
import 'package:flutter/foundation.dart';

// Conditional import for web
import 'csv_export_web.dart' if (dart.library.io) 'csv_export_mobile.dart';

/// Service for exporting data to CSV files
/// Handles CSV generation, formatting, and browser downloads
class CSVExportService {
  /// Export data to CSV file
  /// 
  /// [filename] - Name of the file (without extension)
  /// [headers] - Column headers
  /// [rows] - Data rows (each row is a list matching headers length)
  /// 
  /// Returns the file path on mobile platforms, null on web
  static Future<String?> exportToCSV({
    required String filename,
    required List<String> headers,
    required List<List<dynamic>> rows,
  }) async {
    try {
      // Create CSV data with headers
      final List<List<dynamic>> csvData = [headers, ...rows];

      // Convert to CSV string
      final String csv = const ListToCsvConverter().convert(csvData);

      // Add BOM for proper Excel UTF-8 encoding
      final String csvWithBom = '\uFEFF$csv';

      // Create blob and download (platform-specific)
      final bytes = utf8.encode(csvWithBom);
      
      // Call platform-specific download implementation
      final result = await downloadCSV(filename, bytes);
      
      // On mobile, result is a file path; on web, it's null
      return result is String ? result : null;
    } catch (e) {
      debugPrint('Error exporting CSV: $e');
      rethrow;
    }
  }

  /// Format DateTime to readable string
  static String formatDate(DateTime? date) {
    if (date == null) return 'N/A';
    return DateFormat('MMM dd, yyyy').format(date);
  }

  /// Format DateTime with time
  static String formatDateTime(DateTime? date) {
    if (date == null) return 'N/A';
    return DateFormat('MMM dd, yyyy h:mm a').format(date);
  }

  /// Format timestamp to readable string
  static String formatTimestamp(dynamic timestamp) {
    if (timestamp == null) return 'N/A';
    try {
      if (timestamp is DateTime) {
        return formatDateTime(timestamp);
      }
      // Handle Firestore Timestamp
      final date = timestamp.toDate() as DateTime;
      return formatDateTime(date);
    } catch (e) {
      return 'N/A';
    }
  }

  /// Clean text for CSV (remove commas, quotes, newlines)
  static String cleanText(String? text) {
    if (text == null || text.isEmpty) return '';
    return text
        .replaceAll('\n', ' ')
        .replaceAll('\r', ' ')
        .replaceAll('  ', ' ')
        .trim();
  }

  /// Format phone number
  static String formatPhone(String? phone) {
    if (phone == null || phone.isEmpty) return 'N/A';
    return phone;
  }

  /// Format currency
  static String formatCurrency(double? amount) {
    if (amount == null) return '\$0.00';
    return '\$${amount.toStringAsFixed(2)}';
  }

  /// Format status text
  static String formatStatus(String? status) {
    if (status == null || status.isEmpty) return 'N/A';
    return status.toUpperCase();
  }

  /// Get current timestamp for filename
  static String getTimestamp() {
    return DateFormat('yyyyMMdd_HHmmss').format(DateTime.now());
  }

  /// Generate filename with timestamp
  static String generateFilename(String baseName) {
    return '${baseName}_${getTimestamp()}';
  }

  // ============================================================================
  // DRIVER EXPORT
  // ============================================================================

  /// Export driver list to CSV
  static Future<String?> exportDrivers({
    required List<Map<String, dynamic>> drivers,
    String? filterStatus,
    String? searchQuery,
  }) async {
    final headers = [
      'Name',
      'Email',
      'Phone',
      'Status',
      'Registration Date',
      'Last Active',
      'Total Deliveries',
      'Approved By',
    ];

    final rows = drivers.map((driver) {
      return [
        cleanText(driver['name'] ?? 'N/A'),
        cleanText(driver['email'] ?? 'N/A'),
        formatPhone(driver['phone']),
        formatStatus(driver['approvalStatus'] ?? 'pending'),
        formatTimestamp(driver['createdAt']),
        formatTimestamp(driver['lastActive']),
        (driver['totalDeliveries'] ?? 0).toString(),
        cleanText(driver['approvedBy'] ?? 'N/A'),
      ];
    }).toList();

    // Create filename with filters
    String filename = 'drivers_export';
    if (filterStatus != null && filterStatus != 'all') {
      filename += '_${filterStatus.toLowerCase()}';
    }
    if (searchQuery != null && searchQuery.isNotEmpty) {
      filename += '_filtered';
    }

    return await exportToCSV(
      filename: generateFilename(filename),
      headers: headers,
      rows: rows,
    );
  }

  // ============================================================================
  // DELIVERY EXPORT
  // ============================================================================

  /// Export delivery list to CSV
  static Future<String?> exportDeliveries({
    required List<Map<String, dynamic>> deliveries,
    String? filterStatus,
    String? filterDate,
  }) async {
    final headers = [
      'Tracking Number',
      'Customer Name',
      'Customer Phone',
      'Delivery Address',
      'Scheduled Date',
      'Invoice Date',
      'Status',
      'Driver Name',
      'Notes',
      'Created Date',
      'Completed Date',
    ];

    final rows = deliveries.map((delivery) {
      return [
        cleanText(delivery['trackingNumber'] ?? 'N/A'),
        cleanText(delivery['customerName'] ?? 'N/A'),
        formatPhone(delivery['customerPhone']),
        cleanText(delivery['address'] ?? 'N/A'),
        formatTimestamp(delivery['scheduledDate']),
        formatTimestamp(delivery['invoiceDate']),
        formatStatus(delivery['status'] ?? 'pending'),
        cleanText(delivery['driverName'] ?? 'Unassigned'),
        cleanText(delivery['notes'] ?? ''),
        formatTimestamp(delivery['createdAt']),
        formatTimestamp(delivery['completedAt']),
      ];
    }).toList();

    // Create filename with filters
    String filename = 'deliveries_export';
    if (filterStatus != null && filterStatus != 'all') {
      filename += '_${filterStatus.toLowerCase()}';
    }
    if (filterDate != null && filterDate != 'all') {
      filename += '_${filterDate.toLowerCase()}';
    }

    return await exportToCSV(
      filename: generateFilename(filename),
      headers: headers,
      rows: rows,
    );
  }

  // ============================================================================
  // CLAIM EXPORT
  // ============================================================================

  /// Export claims list to CSV
  static Future<String?> exportClaims({
    required List<Map<String, dynamic>> claims,
    String? filterStatus,
    String? filterType,
  }) async {
    final headers = [
      'Claim ID',
      'Driver Name',
      'Claim Type',
      'Description',
      'Status',
      'Created Date',
      'Updated Date',
      'Resolution Notes',
      'Amount',
    ];

    final rows = claims.map((claim) {
      return [
        cleanText(claim['id'] ?? 'N/A'),
        cleanText(claim['driverName'] ?? 'N/A'),
        cleanText(claim['type'] ?? 'N/A'),
        cleanText(claim['description'] ?? 'N/A'),
        formatStatus(claim['status'] ?? 'pending'),
        formatTimestamp(claim['createdAt']),
        formatTimestamp(claim['updatedAt']),
        cleanText(claim['resolutionNotes'] ?? 'N/A'),
        formatCurrency(claim['claimAmount']?.toDouble()),
      ];
    }).toList();

    // Create filename with filters
    String filename = 'claims_export';
    if (filterStatus != null && filterStatus != 'all') {
      filename += '_${filterStatus.toLowerCase()}';
    }
    if (filterType != null && filterType != 'all') {
      filename += '_${filterType.toLowerCase()}';
    }

    return await exportToCSV(
      filename: generateFilename(filename),
      headers: headers,
      rows: rows,
    );
  }

  // ============================================================================
  // POD EXPORT
  // ============================================================================

  /// Export PODs list to CSV
  static Future<String?> exportPODs({
    required List<Map<String, dynamic>> pods,
    String? filterStatus,
  }) async {
    final headers = [
      'Delivery ID',
      'Tracking Number',
      'Customer Name',
      'Delivery Address',
      'Delivered Date',
      'Status',
      'Has Signature',
      'Has Photo',
      'Recipient Name',
      'Notes',
    ];

    final rows = pods.map((pod) {
      return [
        cleanText(pod['deliveryId'] ?? 'N/A'),
        cleanText(pod['trackingNumber'] ?? 'N/A'),
        cleanText(pod['customerName'] ?? 'N/A'),
        cleanText(pod['deliveryAddress'] ?? 'N/A'),
        formatTimestamp(pod['timestamp']),
        formatStatus(pod['status'] ?? 'delivered'),
        (pod['signatureUrl'] != null && pod['signatureUrl'].toString().isNotEmpty) ? 'Yes' : 'No',
        (pod['photoUrls'] != null && (pod['photoUrls'] as List).isNotEmpty) || (pod['photoUrl'] != null && pod['photoUrl'].toString().isNotEmpty) ? 'Yes' : 'No',
        cleanText(pod['recipientName'] ?? 'N/A'),
        cleanText(pod['notes'] ?? ''),
      ];
    }).toList();

    // Create filename with filters
    String filename = 'pods_export';
    if (filterStatus != null && filterStatus != 'all') {
      filename += '_${filterStatus.toLowerCase()}';
    }

    return await exportToCSV(
      filename: generateFilename(filename),
      headers: headers,
      rows: rows,
    );
  }

  // ============================================================================
  // ANALYTICS EXPORT
  // ============================================================================

  /// Export analytics summary to CSV
  static Future<String?> exportAnalyticsSummary({
    required Map<String, dynamic> summary,
    required DateTime startDate,
    required DateTime endDate,
  }) async {
    final headers = ['Metric', 'Value'];

    final rows = [
      ['Report Period', '${formatDate(startDate)} - ${formatDate(endDate)}'],
      ['Total Deliveries', summary['totalDeliveries']?.toString() ?? '0'],
      ['Completed Deliveries', summary['completedDeliveries']?.toString() ?? '0'],
      ['Pending Deliveries', summary['pendingDeliveries']?.toString() ?? '0'],
      ['In Transit Deliveries', summary['inTransitDeliveries']?.toString() ?? '0'],
      ['Completion Rate', '${summary['completionRate']?.toStringAsFixed(1) ?? '0'}%'],
      ['Active Drivers', summary['activeDrivers']?.toString() ?? '0'],
      ['Total Drivers', summary['totalDrivers']?.toString() ?? '0'],
      ['Total Claims', summary['totalClaims']?.toString() ?? '0'],
      ['Pending Claims', summary['pendingClaims']?.toString() ?? '0'],
      ['Total PODs', summary['totalPODs']?.toString() ?? '0'],
    ];

    return await exportToCSV(
      filename: generateFilename('analytics_summary'),
      headers: headers,
      rows: rows,
    );
  }
}

File: mobile-rn/src/repositories/csvExportService.ts
import Papa from 'papaparse';
import CryptoJS from 'crypto-js';
import Share from 'react-native-share';
import { Vehicle } from '../models/vehicle';
import { CatalogItem, itemCategoryDisplayName } from '../models/catalogItem';

/**
 * Ported from lib/services/csv_export_service.dart + csv_export_mobile.dart (verified
 * against source on 2026-06-22). Pure formatting helpers + CSV generation are 1:1 ports;
 * uses papaparse instead of Dart's `csv` package, matching customerImportService.ts.
 *
 * Deviation: the Dart mobile implementation silently writes the CSV to the device's
 * Downloads/Documents directory via path_provider and returns the file path — there's
 * no share sheet, no user-visible confirmation of *where* the file landed beyond a toast.
 * This port uses react-native-share's native share sheet instead (passing the CSV as a
 * base64 data URL, no temp file needed), which lets the user actually choose where the
 * file goes (Save to Files, share via email, etc.) — strictly more discoverable than a
 * silent background write, and avoids needing react-native-fs for a one-off write.
 * exportToCSV() therefore returns void, not a file path.
 */
export async function exportToCSV(filename: string, headers: string[], rows: unknown[][]): Promise<void> {
  const csv = Papa.unparse([headers, ...rows]);
  const csvWithBom = `﻿${csv}`;
  const base64 = CryptoJS.enc.Base64.stringify(CryptoJS.enc.Utf8.parse(csvWithBom));

  await Share.open({
    url: `data:text/csv;base64,${base64}`,
    filename: `${filename}.csv`,
    type: 'text/csv',
  });
}

/** Mirrors CSVExportService.formatDate(). */
export function formatDate(date?: Date): string {
  if (!date) return 'N/A';
  return date.toLocaleDateString(undefined, { month: 'short', day: '2-digit', year: 'numeric' });
}

/** Mirrors CSVExportService.formatDateTime(). */
export function formatDateTime(date?: Date): string {
  if (!date) return 'N/A';
  const time = date.toLocaleTimeString(undefined, { hour: 'numeric', minute: '2-digit' });
  return `${formatDate(date)} ${time}`;
}

/** Mirrors CSVExportService.cleanText(). */
export function cleanText(text?: string): string {
  if (!text) return '';
  return text.replace(/\n/g, ' ').replace(/\r/g, ' ').replace(/ {2}/g, ' ').trim();
}

/** Mirrors CSVExportService.formatPhone(). */
export function formatPhone(phone?: string): string {
  return phone || 'N/A';
}

/** Mirrors CSVExportService.formatCurrency(). */
export function formatCurrency(amount?: number): string {
  return amount == null ? '$0.00' : `$${amount.toFixed(2)}`;
}

/** Mirrors CSVExportService.formatStatus(). */
export function formatStatus(status?: string): string {
  return status ? status.toUpperCase() : 'N/A';
}

/** Mirrors CSVExportService.getTimestamp(). */
export function getTimestamp(): string {
  const now = new Date();
  const pad = (n: number) => n.toString().padStart(2, '0');
  return `${now.getFullYear()}${pad(now.getMonth() + 1)}${pad(now.getDate())}_${pad(now.getHours())}${pad(now.getMinutes())}${pad(now.getSeconds())}`;
}

/** Mirrors CSVExportService.generateFilename(). */
export function generateFilename(baseName: string): string {
  return `${baseName}_${getTimestamp()}`;
}

// ============================================================================
// DRIVER EXPORT
// ============================================================================

/** Mirrors CSVExportService.exportDrivers(). */
export async function exportDrivers(drivers: Record<string, unknown>[], filterStatus?: string, searchQuery?: string): Promise<void> {
  const headers = ['Name', 'Email', 'Phone', 'Status', 'Registration Date', 'Last Active', 'Total Deliveries', 'Approved By'];

  const rows = drivers.map((driver) => [
    cleanText((driver.name as string) ?? 'N/A'),
    cleanText((driver.email as string) ?? 'N/A'),
    formatPhone(driver.phone as string | undefined),
    formatStatus((driver.approvalStatus as string) ?? 'pending'),
    formatDateTime(driver.createdAt as Date | undefined),
    formatDateTime(driver.lastActive as Date | undefined),
    String(driver.totalDeliveries ?? 0),
    cleanText((driver.approvedBy as string) ?? 'N/A'),
  ]);

  let filename = 'drivers_export';
  if (filterStatus && filterStatus !== 'all') filename += `_${filterStatus.toLowerCase()}`;
  if (searchQuery) filename += '_filtered';

  await exportToCSV(generateFilename(filename), headers, rows);
}

// ============================================================================
// DELIVERY EXPORT
// ============================================================================

/** Mirrors CSVExportService.exportDeliveries(). */
export async function exportDeliveries(deliveries: Record<string, unknown>[], filterStatus?: string, filterDate?: string): Promise<void> {
  const headers = [
    'Tracking Number',
    'Customer Name',
    'Customer Phone',
    'Delivery Address',
    'Scheduled Date',
    'Invoice Date',
    'Status',
    'Driver Name',
    'Notes',
    'Created Date',
    'Completed Date',
  ];

  const rows = deliveries.map((delivery) => [
    cleanText((delivery.trackingNumber as string) ?? 'N/A'),
    cleanText((delivery.customerName as string) ?? 'N/A'),
    formatPhone(delivery.customerPhone as string | undefined),
    cleanText((delivery.address as string) ?? 'N/A'),
    formatDateTime(delivery.scheduledDate as Date | undefined),
    formatDateTime(delivery.invoiceDate as Date | undefined),
    formatStatus((delivery.status as string) ?? 'pending'),
    cleanText((delivery.driverName as string) ?? 'Unassigned'),
    cleanText((delivery.notes as string) ?? ''),
    formatDateTime(delivery.createdAt as Date | undefined),
    formatDateTime(delivery.completedAt as Date | undefined),
  ]);

  let filename = 'deliveries_export';
  if (filterStatus && filterStatus !== 'all') filename += `_${filterStatus.toLowerCase()}`;
  if (filterDate && filterDate !== 'all') filename += `_${filterDate.toLowerCase()}`;

  await exportToCSV(generateFilename(filename), headers, rows);
}

// ============================================================================
// CLAIM EXPORT
// ============================================================================

/** Mirrors CSVExportService.exportClaims(). */
export async function exportClaims(claims: Record<string, unknown>[], filterStatus?: string, filterType?: string): Promise<void> {
  const headers = ['Claim ID', 'Driver Name', 'Claim Type', 'Description', 'Status', 'Created Date', 'Updated Date', 'Resolution Notes', 'Amount'];

  const rows = claims.map((claim) => [
    cleanText((claim.id as string) ?? 'N/A'),
    cleanText((claim.driverName as string) ?? 'N/A'),
    cleanText((claim.type as string) ?? 'N/A'),
    cleanText((claim.description as string) ?? 'N/A'),
    formatStatus((claim.status as string) ?? 'pending'),
    formatDateTime(claim.createdAt as Date | undefined),
    formatDateTime(claim.updatedAt as Date | undefined),
    cleanText((claim.resolutionNotes as string) ?? 'N/A'),
    formatCurrency(claim.claimAmount as number | undefined),
  ]);

  let filename = 'claims_export';
  if (filterStatus && filterStatus !== 'all') filename += `_${filterStatus.toLowerCase()}`;
  if (filterType && filterType !== 'all') filename += `_${filterType.toLowerCase()}`;

  await exportToCSV(generateFilename(filename), headers, rows);
}

// ============================================================================
// POD EXPORT
// ============================================================================

/** Mirrors CSVExportService.exportPODs(). */
export async function exportPODs(pods: Record<string, unknown>[], filterStatus?: string): Promise<void> {
  const headers = [
    'Delivery ID',
    'Tracking Number',
    'Customer Name',
    'Delivery Address',
    'Delivered Date',
    'Status',
    'Has Signature',
    'Has Photo',
    'Recipient Name',
    'Notes',
  ];

  const rows = pods.map((pod) => {
    const photoUrls = pod.photoUrls as string[] | undefined;
    const hasPhoto = (photoUrls?.length ?? 0) > 0 || Boolean(pod.photoUrl);
    return [
      cleanText((pod.deliveryId as string) ?? 'N/A'),
      cleanText((pod.trackingNumber as string) ?? 'N/A'),
      cleanText((pod.customerName as string) ?? 'N/A'),
      cleanText((pod.deliveryAddress as string) ?? 'N/A'),
      formatDateTime(pod.timestamp as Date | undefined),
      formatStatus((pod.status as string) ?? 'delivered'),
      pod.signatureUrl ? 'Yes' : 'No',
      hasPhoto ? 'Yes' : 'No',
      cleanText((pod.recipientName as string) ?? 'N/A'),
      cleanText((pod.notes as string) ?? ''),
    ];
  });

  let filename = 'pods_export';
  if (filterStatus && filterStatus !== 'all') filename += `_${filterStatus.toLowerCase()}`;

  await exportToCSV(generateFilename(filename), headers, rows);
}

// ============================================================================
// ANALYTICS EXPORT
// ============================================================================

/** Mirrors CSVExportService.exportAnalyticsSummary(). */
export async function exportAnalyticsSummary(summary: Record<string, unknown>, startDate: Date, endDate: Date): Promise<void> {
  const headers = ['Metric', 'Value'];

  const rows = [
    ['Report Period', `${formatDate(startDate)} - ${formatDate(endDate)}`],
    ['Total Deliveries', String(summary.totalDeliveries ?? 0)],
    ['Completed Deliveries', String(summary.completedDeliveries ?? 0)],
    ['Pending Deliveries', String(summary.pendingDeliveries ?? 0)],
    ['In Transit Deliveries', String(summary.inTransitDeliveries ?? 0)],
    ['Completion Rate', `${((summary.completionRate as number) ?? 0).toFixed(1)}%`],
    ['Active Drivers', String(summary.activeDrivers ?? 0)],
    ['Total Drivers', String(summary.totalDrivers ?? 0)],
    ['Total Claims', String(summary.totalClaims ?? 0)],
    ['Pending Claims', String(summary.pendingClaims ?? 0)],
    ['Total PODs', String(summary.totalPODs ?? 0)],
  ];

  await exportToCSV(generateFilename('analytics_summary'), headers, rows);
}

// ============================================================================
// VEHICLE EXPORT
// ============================================================================

/**
 * New — no Dart CSVExportService.exportVehicles() ever existed; vehicle_management_desktop.dart
 * built a CSV via a StringBuffer in `_exportToCsv()` that was never written to disk or shared
 * anywhere (dead code — confirmed by direct read, see vault Risk Register). This follows the
 * same per-entity export convention as exportDrivers/exportDeliveries/exportClaims/exportPODs
 * instead of repeating that bug. Takes `Vehicle[]` directly (rather than the other exporters'
 * `Record<string, unknown>[]`) since a typed model already exists here — no need for the loose
 * shape those functions use.
 */
export async function exportVehicles(vehicles: Vehicle[]): Promise<void> {
  const headers = ['Registration', 'Make', 'Model', 'Plate', 'Status', 'Deliveries', 'Last Used', 'Created'];

  const rows = vehicles.map((vehicle) => [
    cleanText(vehicle.registration ?? 'N/A'),
    cleanText(vehicle.make ?? ''),
    cleanText(vehicle.model ?? ''),
    cleanText(vehicle.licensePlate ?? ''),
    formatStatus(vehicle.status ?? 'active'),
    String(vehicle.totalDeliveries ?? 0),
    formatDateTime(vehicle.lastUsedAt),
    formatDateTime(vehicle.createdAt),
  ]);

  await exportToCSV(generateFilename('vehicles_export'), headers, rows);
}

// ============================================================================
// CATALOG ITEM EXPORT
// ============================================================================

/**
 * New — no Dart CSVExportService.exportCatalogItems() ever existed; item_catalog_desktop.dart's
 * `_exportToCsv()` builds a CSV StringBuffer and then only shows a `SnackBar` ("CSV ready: N
 * items") — the buffer is never written to disk or shared anywhere (dead code, same pattern as
 * vehicle_management_desktop's bug — see vault Risk Register). Wired for real here.
 */
export async function exportCatalogItems(items: CatalogItem[]): Promise<void> {
  const headers = ['Description', 'SKU', 'Category', 'Unit', 'Default Qty', 'Price', 'Usage Count', 'Created'];

  const rows = items.map((item) => [
    cleanText(item.description ?? 'N/A'),
    cleanText(item.sku ?? ''),
    cleanText(itemCategoryDisplayName(item.category)),
    cleanText(item.unit ?? ''),
    item.defaultQuantity != null ? String(item.defaultQuantity) : '',
    item.unitPrice != null ? item.unitPrice.toFixed(2) : '',
    String(item.usageCount ?? 0),
    formatDateTime(item.createdAt),
  ]);

  await exportToCSV(generateFilename('catalog_items_export'), headers, rows);
}
